portal-impl/src/com/liferay/portal/util/HttpImpl.java

```diff
@@ -1444,6 +1449,69 @@ protected CloseableHttpClient getCloseableHttpClient(HttpHost proxyHost) {
 		return _closeableHttpClient;
 	}
 
+	protected int getMaxURLDepth(
+		String url, int curLength, int maxLength, int count) {
+
+		String ampersand = StringPool.AMPERSAND;
+		String equal = StringPool.EQUAL;
+		String question = StringPool.QUESTION;
+
+		for (int i = 0; i < count; i++) {
+			ampersand = URLCodec.encodeURL(ampersand);
+			equal = URLCodec.encodeURL(equal);
+			question = URLCodec.encodeURL(question);
+		}
+
+		url = StringUtil.replace(url, question, ampersand);
+
+		List<String> redirectParams = new ArrayList<>();
+
+		String[] params = StringUtil.split(url, ampersand);
+
+		for (int i = 0; i < params.length; i++) {
+			if (params[i].contains("_backURL" + equal) ||
+				params[i].contains("_redirect" + equal) ||
+				params[i].contains("_returnToFullPageURL" + equal) ||
+				params[i].startsWith("redirect")) {
+
+				redirectParams.add(params[i]);
+			}
+			else {
+				curLength += params[i].length();
+
+				if (params[i].contains(equal)) {
+					curLength += ampersand.length();
+				}
+			}
+		}
+
+		if (curLength > maxLength) {
+			return count - 1;
+		}
+
+		StringBundler sb = new StringBundler();
+
+		for (String redirectParam : redirectParams) {
+			int pos = redirectParam.indexOf(equal) + equal.length();
+
+			curLength += pos;
+
+			curLength += ampersand.length();
+
+			sb.append(redirectParam.substring(pos));
+			sb.append(URLCodec.encodeURL(ampersand));
+		}
+
+		if ((sb.index() <= 0) || (curLength > maxLength)) {
```

Comment:
... and if curLength now exceeds the limit, we know we cannot go any deeper, so return the current count.  Or, if we did not have any nestable params, we have hit the end of our URL without reaching the limit, so return the current count as it is the depth of the URL.